kasa/klapprotocol.py

```
@@ -0,0 +1,512 @@
+            TPLinkKlap.TP_SESSION_COOKIE_NAME
+        )
+        session.cookie_jar.clear()
+        session.cookie_jar.update_cookies(
+            {TPLinkKlap.TP_SESSION_COOKIE_NAME: cookie}, URL(url)
+        )
+
+    def clear_cookies(self, session):
+        """Clear out all cookies for new handshake."""
+        session.cookie_jar.clear()
+        self.jar.clear()
+
+    async def perform_handshake1(
+        self, session, new_local_seed: Optional[bytes] = None
+    ) -> None:
+        """Perform handshake1.  Resets authentication_failed to False at the start."""
+        self.authentication_failed = False
+
+        self.clear_cookies(session)
+
+        if new_local_seed is not None:
+            self._local_seed = new_local_seed
+        else:
+            self._local_seed = self.get_local_seed()
+
+        # Handshake 1 has a payload of local_seed
+        # and a response of 16 bytes, followed by sha256(clientBytes | authenticator)
+        self.handshake_done = False
+
+        payload = self._local_seed
+
+        url = f"http://{self.host}/app/handshake1"
+
+        response_status, response_data = await self.session_post(
+            session, url, data=payload
+        )
+
+        cookie = self.jar.filter_cookies(url).get(self.TP_SESSION_COOKIE_NAME)
+
+        if response_status != 200:
+            raise SmartDeviceAuthenticationException(
+                "Device %s responded with %d to handshake1, this is probably not a klap device"
+                % (self.host, response_status)
+            )
+        self.handle_cookies(session, url)
+
+        self.remote_seed = response_data[0:16]
+        server_hash = response_data[16:]
+
+        cookie = self.jar.filter_cookies(url).get(self.TP_SESSION_COOKIE_NAME)
+        _LOGGER.debug(
+            f"Handshake1 posted at {datetime.datetime.now()}.  Host is {self.host}, Session cookie is {cookie}, Response status is {response_status}, Request was {self.local_auth_hash.hex()}"
+        )
```

Does the device set the cookie always? The same cookie name is used by tapo devices, but as the request on P110 did not return a cookie, maybe this should raise an early exception if no cookie is being set?

---

As far as I am aware the kasa klap devices always set the cookie after handshake1.